### ISSUE TITLE
Allow for one sided relations (the reverse field on related model is not created).

### ISSUE THREAD

**Issue by Erhannis** [collerek/ormar#118]
**Describe the bug**
When I return a list of objects, from a FastAPI endpoint, unless I have all the fields loaded I get a `ReferenceError: weakly-referenced object no longer exists` - even if I've excluded those fields from the result.  (It's possible this is more of a bug on FastAPI's side.)

**To Reproduce**
In an endpoint, load an object with a virtual list of foreign objects.  Return the list from the endpoint.  A `ReferenceError` occurs.  However, if you have called `.load()` on each element of the list, the error does not occur.  That would be fine, but lacking a `.load_all()` method, it's annoying (and inefficient) to do all those `.load()`s - especially when you don't actually want that field, anyway.  See following code:

```
class OtherThing(ormar.Model):
    class Meta(BaseMeta):
        tablename = "other_things"
    id: UUID = ormar.UUID(primary_key=True, default=uuid4)
    name: str = ormar.Text(default="")
    ot_contents: str = ormar.Text(default="")
 
class Thing(ormar.Model):
    class Meta(BaseMeta):
        tablename = "things"
    id: UUID = ormar.UUID(primary_key=True, default=uuid4)
    name: str = ormar.Text(default="")
    js: pydantic.Json = ormar.JSON(nullable=True)
    other_thing: Optional[OtherThing] = ormar.ForeignKey(OtherThing, nullable=True)

@app.post("/test/1")
async def post_test_1():
    ot = OtherThing()
    ot.ot_contents = "otc"
    t1 = Thing(other_thing=ot)
    t1.name = "t1"
    t2 = Thing(other_thing=ot)
    t2.name = "t2"
    t3 = Thing(other_thing=ot)
    t3.name = "t3"
    await ot.save()
    await t1.save()
    await t2.save()
    await t3.save()

@app.get("/test/2", response_model=List[Thing], response_model_exclude={"other_thing"})
async def get_test_2():
    ots = await OtherThing.objects.all()
    ot = ots[0]
    ts = await ot.things.all()
    #for t in ts:
    #    await t.load()
    return ts
```
(Note that uncommenting the load lines makes the error not occur.)

And here's a log:
```
[32mINFO[0m:     127.0.0.1:62163 - "[1mPOST /test/1 HTTP/1.1[0m" [32m200 OK[0m
[32mINFO[0m:     127.0.0.1:62168 - "[1mGET /test/2 HTTP/1.1[0m" [91m500 Internal Server Error[0m
[31mERROR[0m:    Exception in ASGI application
Traceback (most recent call last):
  File "c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\uvicorn\protocols\http\h11_impl.py", line 394, in run_asgi
    result = await app(self.scope, self.receive, self.send)
  File "c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\uvicorn\middleware\proxy_headers.py", line 45, in __call__
    return await self.app(scope, receive, send)
  File "c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\fastapi\applications.py", line 199, in __call__
    await super().__call__(scope, receive, send)
  File "c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\starlette\applications.py", line 111, in __call__
    await self.middleware_stack(scope, receive, send)
  File "c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\starlette\middleware\errors.py", line 181, in __call__
    raise exc from None
  File "c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\starlette\middleware\errors.py", line 159, in __call__
    await self.app(scope, receive, _send)
  File "c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\starlette\exceptions.py", line 82, in __call__
    raise exc from None
  File "c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\starlette\exceptions.py", line 71, in __call__
    await self.app(scope, receive, sender)
  File "c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\starlette\routing.py", line 566, in __call__
    await route.handle(scope, receive, send)
  File "c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\starlette\routing.py", line 227, in handle
    await self.app(scope, receive, send)
  File "c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\starlette\routing.py", line 41, in app
    response = await func(request)
  File "c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\fastapi\routing.py", line 209, in app
    response_data = await serialize_response(
  File "c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\fastapi\routing.py", line 137, in serialize_response
    return jsonable_encoder(response_content)
  File "c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\fastapi\encoders.py", line 104, in jsonable_encoder
    jsonable_encoder(
  File "c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\fastapi\encoders.py", line 47, in jsonable_encoder
    obj_dict = obj.dict(
  File "c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\ormar\models\newbasemodel.py", line 633, in dict
    dict_instance = self._extract_nested_models(
  File "c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\ormar\models\newbasemodel.py", line 569, in _extract_nested_models
    dict_instance[field] = nested_model.dict(
ReferenceError: weakly-referenced object no longer exists
ERROR:uvicorn.error:Exception in ASGI application
Traceback (most recent call last):
  File "c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\uvicorn\protocols\http\h11_impl.py", line 394, in run_asgi
    result = await app(self.scope, self.receive, self.send)
  File "c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\uvicorn\middleware\proxy_headers.py", line 45, in __call__
    return await self.app(scope, receive, send)
  File "c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\fastapi\applications.py", line 199, in __call__
    await super().__call__(scope, receive, send)
  File "c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\starlette\applications.py", line 111, in __call__
    await self.middleware_stack(scope, receive, send)
  File "c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\starlette\middleware\errors.py", line 181, in __call__
    raise exc from None
  File "c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\starlette\middleware\errors.py", line 159, in __call__
    await self.app(scope, receive, _send)
  File "c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\starlette\exceptions.py", line 82, in __call__
    raise exc from None
  File "c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\starlette\exceptions.py", line 71, in __call__
    await self.app(scope, receive, sender)
  File "c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\starlette\routing.py", line 566, in __call__
    await route.handle(scope, receive, send)
  File "c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\starlette\routing.py", line 227, in handle
    await self.app(scope, receive, send)
  File "c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\starlette\routing.py", line 41, in app
    response = await func(request)
  File "c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\fastapi\routing.py", line 209, in app
    response_data = await serialize_response(
  File "c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\fastapi\routing.py", line 137, in serialize_response
    return jsonable_encoder(response_content)
  File "c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\fastapi\encoders.py", line 104, in jsonable_encoder
    jsonable_encoder(
  File "c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\fastapi\encoders.py", line 47, in jsonable_encoder
    obj_dict = obj.dict(
  File "c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\ormar\models\newbasemodel.py", line 633, in dict
    dict_instance = self._extract_nested_models(
  File "c:\users\USER\appdata\local\programs\python\python38\lib\site-packages\ormar\models\newbasemodel.py", line 569, in _extract_nested_models
    dict_instance[field] = nested_model.dict(
ReferenceError: weakly-referenced object no longer exists
```

**Expected behavior**
No error when the field in question has been excluded.  (Possibly no error in general, but I'm waffly on that.)

**Screenshots**
~~If applicable, add screenshots to help explain your problem.~~

**Versions (please complete the following information):**
 - Database backend used: sqlite
 - Python version: 3.8.0
 - `ormar` version: 0.9.4
 - `pydantic` version: 1.7.3
 - `fastapi` version: 0.63.0

**Additional context**
~~Add any other context about the problem here.~~

**Comment by collerek** [comment]
That's not really a bug it's a feature (yeah I know how that sounds :D)

But because right now all relations are bidirectional. 
So when you have relation `A <=> B`: 

* `A` has relation to `B`
* `B` has relation to `A`

So in order to avoid circular references, only the active side of the query is keeping the concrete model reference, the other side is keeping a week reference so it's easier to garbage collect the whole thing. 

The active relation is the one that you come from -> so if you query from `A` to `B` , `A` keeps a reference to `B`, and `B` has a week reference to `A`, when you query from `B` to `A` it's the opposite.

Since you query from `OtherThing` to `Things`, `Things` have week reference to `OtherThing` (`ot` in your sample code), and since you discard this value (return only `ts`) the `ot` is garbage collected by the time `fastapi` is handling it and the weakref is broken.

Now I am willing to change both sides to concrete models instead of weakrefs, cause I know that python can handle circular references and will garbage collect them, but it's using a different algorithm to determine if the loop A <=> B has no external references and it comes with a performance hit. When you think about A <=> B loop it might look simple and not costly,
but when you look at a complicated `ormar` query like this:
```
        C - D - E ... - Z
      /
    B - Q -T
  /
A        O - P
    \   /
      X 
        \
          Y  - Z.... - F
```

It would create a huge loop of references from A - Z and back. That's why I would have to properly check performance and memory leaks in that scenario, which is not trivial (would have to run with thousands of models for quite some time and not with toy examples like in current tests), and since I didn't have the time to do so for now it stays as it is.

**Now that does not mean you cannot do anything about this.**

BTW. I know that it's your simple tests etc. but you are using the db operations very inefficiently -> you should avoid issuing the additional queries if you can avoid them as those are most costly operations in general while using `ormar` (and any orm). You might already know that, but I want to point this out for other people that might read this.

As for your options I embedded the tips/answers in the code. I included also some remarks/comments.

```python
from typing import List, Optional
from uuid import UUID, uuid4

import databases
import pydantic
import pytest
import sqlalchemy
from fastapi import FastAPI
from starlette.testclient import TestClient

import ormar
from tests.settings import DATABASE_URL

app = FastAPI()

database = databases.Database(DATABASE_URL, force_rollback=True)
metadata = sqlalchemy.MetaData()

app.state.database = database


@app.on_event("startup")
async def startup() -> None:
    database_ = app.state.database
    if not database_.is_connected:
        await database_.connect()


@app.on_event("shutdown")
async def shutdown() -> None:
    database_ = app.state.database
    if database_.is_connected:
        await database_.disconnect()


@pytest.fixture(autouse=True, scope="module")
def create_test_database():
    engine = sqlalchemy.create_engine(DATABASE_URL)
    metadata.create_all(engine)
    yield
    metadata.drop_all(engine)


class BaseMeta(ormar.ModelMeta):
    database = database
    metadata = metadata


class OtherThing(ormar.Model):
    class Meta(BaseMeta):
        tablename = "other_things"

    id: UUID = ormar.UUID(primary_key=True, default=uuid4)
    name: str = ormar.Text(default="")
    ot_contents: str = ormar.Text(default="")


class Thing(ormar.Model):
    class Meta(BaseMeta):
        tablename = "things"

    id: UUID = ormar.UUID(primary_key=True, default=uuid4)
    name: str = ormar.Text(default="")
    js: pydantic.Json = ormar.JSON(nullable=True)
    other_thing: Optional[OtherThing] = ormar.ForeignKey(OtherThing, nullable=True)


@app.post("/test/1")
async def post_test_1():
    # don't split initialization and attribute assignment
    ot = await OtherThing(ot_contents="otc").save()
    await Thing(other_thing=ot, name='t1').save()
    await Thing(other_thing=ot, name="t2").save()
    await Thing(other_thing=ot, name="t3").save()

    # if you do not care about returned object you can even go with bulk_create
    # all of them are created in one transaction
    # things = [Thing(other_thing=ot, name='t1'),
    #           Thing(other_thing=ot, name="t2"),
    #           Thing(other_thing=ot, name="t3")]
    # await Thing.objects.bulk_create(things)


# OPTION 1
@app.get("/test/2", response_model=List[Thing])
async def get_test_2():
    # if you only query for one use get or first
    ot = await OtherThing.objects.get()
    ts = await ot.things.all()
    # specifically null out the relation on things before return
    for t in ts:
        t.remove(ot, name='other_thing')
    return ts


# OPTION 2
@app.get("/test/3", response_model=List[Thing])
async def get_test_3():
    ot = await OtherThing.objects.select_related('things').get()
    # exclude unwanted field while ot is still in scope
    # in order not to pass it to fastapi
    return [t.dict(exclude={'other_thing'}) for t in ot.things]

# OPTION 3
@app.get("/test/4", response_model=List[Thing], response_model_exclude={"other_thing"})
async def get_test_4():
    ot = await OtherThing.objects.get()
    # query from the active side
    return await Thing.objects.all(other_thing=ot)


@app.get("/get_ot/", response_model=OtherThing)
async def get_ot():
    return await OtherThing.objects.get()

# OPTION 4
# more real life (usually) is not getting some random OT and get it's Things
# but query for a specific one by some kind of id
@app.get("/test/5/{thing_id}", response_model=List[Thing],
         response_model_exclude={"other_thing"})
async def get_test_5(thing_id: UUID):
    return await Thing.objects.all(other_thing__id=thing_id)


def test_endpoints():
    client = TestClient(app)
    with client:
        resp = client.post('/test/1')
        assert resp.status_code == 200

        resp2 = client.get('/test/2')
        assert resp2.status_code == 200
        assert len(resp2.json()) == 3

        resp3 = client.get('/test/3')
        assert resp3.status_code == 200
        assert len(resp3.json()) == 3

        resp4 = client.get('/test/4')
        assert resp4.status_code == 200
        assert len(resp4.json()) == 3

        ot = OtherThing(**client.get('/get_ot/').json())
        resp5 = client.get(f"/test/5/{ot.id}")
        assert resp5.status_code == 200
        assert len(resp5.json()) == 3

```

**Note that this will no longer apply if the one sided relation will be allowed (`skip_reverse` param on relation fields), as there would be no back reference from child models back to main models.** 

That's why instead of closing this I will change this into enhancement.

**Comment by Erhannis** [comment]
Oh; yeah, I expected the weak references were there to avoid reference cycles.  That's fine, if not perfect from a user's viewpoint.  My objection was that it causes an error, even though I've excluded that field.  It feels like a peculiarity of Ormar that triggers an arguable bug in FastAPI.  Perhaps this would be something to file with FastAPI?

Options 2 and 4 are useful, though; I'll probably go with option 4, since that IS my use case, it's just as simple as the other ways, and is (presumably) more efficient.  (Though, often, "fast enough" is sufficient; I'll often opt for "simple" over "fast".)

**Comment by collerek** [comment]
Hmm when i think about it again, maybe it is a bug in ormar. I should discard the expired refs when extracting nested models. Cause fastapi expects BaseModel from pydantic and i think in those there are no weakrefs. The problem is that fastapi instantiate models in order to validate them, dump to json etc., and only at the end the field is excluded. Will look into this regardless of the enhancement.

**Comment by collerek** [comment]
Fixed (bug) and added (skip_reverse) in 0.10.3
